Add a pass to collect dropped var stats for MIR.

rastogishubham opened the pull request:
This patch uses the DroppedVariableStats class to add dropped variable
statistics for MIR passes.

rastogishubham commented:
This is a part of a stack of patches, with https://github.com/llvm/llvm-project/pull/115563 being the first one.

When I run the MIR stats on SemaExpr.cpp.o with the command line:

```
clang++ -g -O3 -mllvm -dropped-variable-stats-mir -DGTEST_HAS_RTTI=0 -D_DEBUG -D_GLIBCXX_ASSERTIONS -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/Users/shubham/Development/llvm-project/build_ninja/tools/clang/lib/Sema -I/Users/shubham/Development/llvm-project/clang/lib/Sema -I/Users/shubham/Development/llvm-project/clang/include -I/Users/shubham/Development/llvm-project/build_ninja/tools/clang/include -I/Users/shubham/Development/llvm-project/build_ninja/include -I/Users/shubham/Development/llvm-project/llvm/include -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -fno-common -Woverloaded-virtual -Wno-nested-anon-types -DNDEBUG -std=c++17 -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX15.0.sdk  -fno-exceptions -funwind-tables -fno-rtti -UNDEBUG -MD -MT tools/clang/lib/Sema/CMakeFiles/obj.clangSema.dir/SemaExpr.cpp.o -MF tools/clang/lib/Sema/CMakeFiles/obj.clangSema.dir/SemaExpr.cpp.o.d -o tools/clang/lib/Sema/CMakeFiles/obj.clangSema.dir/SemaExpr.cpp.o -c /Users/shubham/Development/llvm-project/clang/lib/Sema/SemaExpr.cpp
```

I get the results:

'Early Tail Duplication', 12
'Branch Probability Basic Block Placement', 48
'Control Flow Optimizer', 152

rastogishubham commented:
@OCHyams Done!

OCHyams commented on llvm/include/llvm/CodeGen/DroppedVariableStats.h:
```diff
@@ -0,0 +1,199 @@
+
+/// A class to collect and print dropped debug information due to MIR
+/// optimization passes. After every MIR pass is run, it will print how many
+/// #DBG_VALUEs were dropped due to that pass.
```
```suggestion
/// DBG_VALUEs were dropped due to that pass.
```
nit

OCHyams replied:
this typo still seems to exist? (maybe the change got lost in a force push?)

OCHyams reviewed (commented):
What's actually different between the MIR and IR versions. Skimming the patch it looks like:

* the types `Function` -> `MachineFunction`, `Instruction` -> `MachineInstr` for iterating
* getting VarID from `DbgVaraibleRecord` -> `MachineInstr`

I feel like this could probably be re-structured so that most of the machinery between the IR and MIR versions is shared with some amount of templating or/or generic lambda usage, to reduce code duplication. Wdyt?

rastogishubham committed:
Add a pass to collect dropped var stats for MIR.

This patch extends the DroppedVariableStats class to add dropped
variable statistics for MIR passes.

OCHyams reviewed (approved):
Just reviewing the 3rd commit (other commit review comments over on #115563), LGTM (with one tiny inline nit to resolve).

rastogishubham commented:
Had to revert with 81924ac1fb63273fc4648029301869a085bb9dac for the buildbot failures

nikic commented on llvm/include/llvm/CodeGen/MachineFunctionPass.h:
```diff
@@ -18,6 +18,7 @@
 #ifndef LLVM_CODEGEN_MACHINEFUNCTIONPASS_H
 #define LLVM_CODEGEN_MACHINEFUNCTIONPASS_H
 
+#include "llvm/CodeGen/DroppedVariableStats.h"
```
This include is mindbogglingly expensive (2% on a single-backend build). Please change your design in a way that avoids pulling in the whole kitchen sink into MachineFunctionPass.h...
